File: IN/KinectControllerCore/KinectController.cs
using Microsoft.Kinect;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Kinect.Toolbox;
using Kinect.Toolbox.Record;
using System.Globalization;
using System.Configuration;
using Microsoft.Speech.Recognition;
using Microsoft.Speech.AudioFormat;
using Microsoft.Kinect.Toolkit.FaceTracking;

namespace KinectControllerCore
{
    public class KinectController : IKinectController
    {
        private static List<KinectController> controllers = new List<KinectController>();
        readonly ContextTracker contextTracker = new ContextTracker();

        private List<string> lines = new List<string>();
        private int counter = 0;

        private KinectController() { }

        private KinectSensor Sensor { get; set; }

        private SpeechRecognitionEngine SpeechRecognizer { get; set; }

        private FaceTracker HeadTracker { get; set; }

        private Skeleton[] skeletons;

        public static int NewInstance()
        {
            var controller = new KinectController();
            controllers.Add(controller);
            return controllers.Count - 1;
        }

        public static KinectController GetInstance(int index)
        {
            if (!controllers.Any())
                return null;
            return controllers[index];
        }

        public event EventHandler<GestureDetectedEventArgs> GestureDetected;

        public bool Initialize()
        {
            foreach (var sensor in KinectSensor.KinectSensors)
            {
                if (sensor.Status == KinectStatus.Connected)
                {
                    Sensor = sensor;
                    break;
                }
            }
            if (Sensor == null)
                return false;
            InitSensor();
            return true;
        }

        private void InitSensor()
        {
            Sensor.SkeletonStream.Enable(new TransformSmoothParameters
            {
                Smoothing = 0.5f,
                Correction = 0.5f,
                Prediction = 0.5f,
                JitterRadius = 0.05f,
                MaxDeviationRadius = 0.04f
            });
            Sensor.ColorStream.Enable();
            Sensor.DepthStream.Enable();
            Sensor.AllFramesReady += OnAllFramesReady;
            EnsureHeadTracker();
            Sensor.Start();

            SpeechRecognizer = new SpeechRecognitionEngine(new CultureInfo("en-US"));
            GrammarBuilder grammarBuilder = new GrammarBuilder();
            grammarBuilder.Culture = new CultureInfo("en-US");
            var commands = new Choices();
            commands.Add("open");
            commands.Add("close");
            grammarBuilder.Append(commands);
            SpeechRecognizer.LoadGrammar(new Grammar(grammarBuilder));
            SpeechRecognizer.SpeechRecognized += SpeechRecognizedHandler;
            SpeechRecognizer.SetInputToAudioStream(Sensor.AudioSource.Start(),
                new SpeechAudioFormatInfo(EncodingFormat.Pcm, 16000, 16, 1, 32000, 2, null));
            SpeechRecognizer.RecognizeAsync(RecognizeMode.Multiple);
        }

        void OnGestureDetected(string obj, double diff)
        {
            GestureType gesture = GestureType.Unknown;
            Enum.TryParse(obj, out gesture);
            if (GestureDetected != null)
                GestureDetected(this, new GestureDetectedEventArgs()
                {
                    Gesture = gesture,
                    Diff = diff
                });
        }

        private bool EnsureHeadTracker()
        {
            if (HeadTracker==null)
            {
                try
                {
                    HeadTracker = new FaceTracker(Sensor);
                }
                catch (InvalidOperationException)
                {
                    return false;
                }
            }
            return true;
        }

        void OnAllFramesReady(object sender, AllFramesReadyEventArgs e)
        {
            using (SkeletonFrame skeletonFrame = e.OpenSkeletonFrame())
            {
                using (var colorFrame = e.OpenColorImageFrame())
                {
                    using (var depthFrame = e.OpenDepthImageFrame())
                    {

                        if (skeletonFrame == null)
                            return;

                        skeletonFrame.GetSkeletons(ref skeletons);

                        if (!skeletons.All(s => s.TrackingState == SkeletonTrackingState.NotTracked))
                            ProcessFrame(colorFrame, depthFrame, skeletonFrame);
                    }
                }
            }
        }

        void SpeechRecognizedHandler(object sender, SpeechRecognizedEventArgs e)
        {
            if (SpeechRecognized != null)
                SpeechRecognized(this, new VoiceCommandEventArgs()
                {
                    Command = e.Result.Text
                });
        }
        
        public void Dispose()
        {
            if (Sensor != null)
                Sensor.Dispose();
            if (SpeechRecognizer != null)
                SpeechRecognizer.Dispose();
            SpeechRecognizer = null;
            Sensor = null;

            skeletons = null;

            foreach (var item in SpeechRecognized.GetInvocationList())
                SpeechRecognized -= (EventHandler<VoiceCommandEventArgs>)item;
            foreach (var item in GestureDetected.GetInvocationList())
                GestureDetected -= (EventHandler<GestureDetectedEventArgs>)item;
        }

        private void ProcessFrame(ColorImageFrame colorFrame, DepthImageFrame depthFrame, ReplaySkeletonFrame skeletonFrame)
        {
            Dictionary<int, string> stabilities = new Dictionary<int, string>();
            foreach (var skeleton in skeletonFrame.Skeletons)
            {
                if (skeleton.TrackingState != SkeletonTrackingState.Tracked)
                    continue;

                contextTracker.Add(skeleton.Position.ToVector3(), skeleton.TrackingId);
                stabilities.Add(skeleton.TrackingId, contextTracker.IsStableRelativeToCurrentSpeed(skeleton.TrackingId) ? "Stable" : "Non stable");
                if (!contextTracker.IsStableRelativeToCurrentSpeed(skeleton.TrackingId))
                    continue;

                SkeletonPoint center = skeleton.Position;
                foreach (Joint joint in skeleton.Joints)
                {
                    if (joint.JointType == JointType.HandLeft)
                    {
                        if (center.Z - joint.Position.Z > 0.25)
                            OnGestureDetected(GestureType.LeftInFront.ToString(), center.Z - joint.Position.Z);
                    }
                    else if (joint.JointType == JointType.HandRight)
                    {
                        if (center.Z - joint.Position.Z > 0.25)
                            OnGestureDetected(GestureType.RightInFront.ToString(), center.Z - joint.Position.Z);
                    }
                }

                if (EnsureHeadTracker())
                {
                    byte[] colorImage = new byte[colorFrame.PixelDataLength];
                    colorFrame.CopyPixelDataTo(colorImage);
                    short[] depthImage = new short[depthFrame.PixelDataLength];
                    depthFrame.CopyPixelDataTo(depthImage);
                    var result = HeadTracker.Track(colorFrame.Format, colorImage, depthFrame.Format, depthImage, skeleton);
                    if (result.TrackSuccessful)
                    {
                        if (FaceTracked != null)
                            FaceTracked(this, new FaceTrackedEventArgs()
                            {
                                Pitch = result.Rotation.X,
                                Roll = result.Rotation.Z,
                                Yaw = result.Rotation.Y
                            });
                    }
                }
            }
        }

        public event EventHandler<VoiceCommandEventArgs> SpeechRecognized;

        public event EventHandler<FaceTrackedEventArgs> FaceTracked;
    }
}

File: IN/KinectController/KinectController.cpp
// This is the main DLL file.

#include "stdafx.h"

namespace KinectController
{
	ref class GestureHandler
	{
	public:
		KinectController* target;
		void OnGestureDetected(System::Object ^sender, KinectControllerCore::GestureDetectedEventArgs ^e);
	};

	ref class SpeechHandler
	{
	public:
		KinectController* target;
		void OnSpeechRecognized(System::Object ^sender, KinectControllerCore::VoiceCommandEventArgs ^e);
	};

	ref class HeadHandler
	{
	public:
		KinectController* target;
		void OnHeadTracked(System::Object ^sender, KinectControllerCore::FaceTrackedEventArgs ^e);
	};

	KinectController::KinectController()
	{
		m_instance = KinectControllerCore::KinectController::NewInstance();

		GestureHandler^ gestureHandler = gcnew GestureHandler();
		gestureHandler->target = this;
		KinectControllerCore::KinectController::GetInstance(m_instance)->GestureDetected += gcnew System::EventHandler<KinectControllerCore::GestureDetectedEventArgs ^>(gestureHandler, &GestureHandler::OnGestureDetected);

		SpeechHandler^ speechHandler = gcnew SpeechHandler();
		speechHandler->target = this;
		KinectControllerCore::KinectController::GetInstance(m_instance)->SpeechRecognized += gcnew System::EventHandler<KinectControllerCore::VoiceCommandEventArgs ^>(speechHandler, &SpeechHandler::OnSpeechRecognized);

		HeadHandler^ headHandler = gcnew HeadHandler();
		headHandler->target = this;
		KinectControllerCore::KinectController::GetInstance(m_instance)->FaceTracked += gcnew System::EventHandler<KinectControllerCore::FaceTrackedEventArgs ^>(headHandler, &HeadHandler::OnHeadTracked);
	}

	KinectController::~KinectController()
	{
		KinectControllerCore::KinectController^ intController = KinectControllerCore::KinectController::GetInstance(m_instance);
		if (intController != nullptr)
			intController->~KinectController();
	}

	bool KinectController::initialize()
	{
		KinectControllerCore::KinectController^ intController = KinectControllerCore::KinectController::GetInstance(m_instance);
		bool result = false;
		try
		{
			result = intController->Initialize();
		}
		catch (std::exception exc)
		{
			printf("%s\n", exc.what());
		}
		catch (System::Exception^ exc)
		{
			printf("%s\n", cstr(exc->Message).c_str());
		}
		return result;
	}

	void KinectController::dispose()
	{
		KinectControllerCore::KinectController^ intController = KinectControllerCore::KinectController::GetInstance(m_instance);
		intController->~KinectController();
	}

	void KinectController::setGestureDetectedHandler(std::function<void(GestureType gesture, double diff)> gestureDetectedHandler)
	{
		this->gestureDetectedHandler = gestureDetectedHandler;
	}

	std::function<void(GestureType gesture, double diff)> KinectController::getGestureDetectedHandler()
	{
		return gestureDetectedHandler;
	}

	void KinectController::setSpeechRecognizedHandler(std::function<void(std::string)> handler)
	{
		this->speechRecognizedHandler = handler;
	}

	std::function<void(std::string)> KinectController::getSpeechRecognizedHandler()
	{
		return speechRecognizedHandler;
	}

	void KinectController::setHeadTrackedHandler(std::function<void(float p, float r, float y)> handler)
	{
		this->headTrackedHandler = handler;
	}
	std::function<void(float, float, float)> KinectController::getHeadTrackedHandler()
	{
		return this->headTrackedHandler;
	}

	void GestureHandler::OnGestureDetected(System::Object ^sender, KinectControllerCore::GestureDetectedEventArgs ^e)
	{
		if (target->getGestureDetectedHandler())
			target->getGestureDetectedHandler()((GestureType)e->Gesture, e->Diff);
	}

	void SpeechHandler::OnSpeechRecognized(System::Object ^sender, KinectControllerCore::VoiceCommandEventArgs ^e)
	{
		if (target->getSpeechRecognizedHandler())
			target->getSpeechRecognizedHandler()(cstr(e->Command));
	}

	void HeadHandler::OnHeadTracked(System::Object ^sender, KinectControllerCore::FaceTrackedEventArgs ^e)
	{
		if (target->getHeadTrackedHandler())
			target->getHeadTrackedHandler()(e->Pitch, e->Roll, e->Yaw);
	}
}